### PR TITLE
Split Gulpfiles

### PR DESCRIPTION
**Pull request by bbesmanoff**
This puts us down a path where the server and app can behave like independent projects

### CODE REVIEW

File: Gulpfile.js [hunk 1]
```diff
@@ -14,110 +15,9 @@ var lint = require('gulp-tslint');
...
-  return gulp.src(coverageFile)
-    .pipe(istanbulReport({
-      reporters: ['text-summary', 'html']
-    }));
-});
-
-gulp.task('test', gulp.series('test-run', 'test-coverage-report'));
+gulp.task('lint', gulp.parallel('lint:app'));
+gulp.task('build', gulp.parallel('build:app'));
+gulp.task('test', gulp.parallel('test:app'));
```
Comment (djrenren) [inline Gulpfile.js:9]: Shouldn't these be the other way around? I'd expect the Gulpfile for app to just say `lint`, `build`, or `test` and the overarching Gulpfile to say `build:app`

File: .gitignore [hunk 2]
```diff
@@ -31,6 +31,8 @@ node_modules
 
 !typings
 app/**/*.js
```
Comment (djrenren) [inline .gitignore]: If we're compiling to tmp, then we should change this ignore. We should just ignore everything under `tmp/` and `test/instrumented` that way we don't have to have explicit allows for the js files we're using.

File: .gitignore [hunk 3]
```diff
@@ -31,6 +31,8 @@ node_modules
...
+!app/gulpfile.js
+!app/test/url_es3.js
 test/app/*.js
```
Comment (djrenren) [inline .gitignore]: These are defunct

File: app/gulpfile.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+gulp.task('build:app', gulp.series(function() {
+  var project = ts.createProject(_localFile('tsconfig.json'), {typescript: typescript});
+
+  return project.src()
+    .pipe(sourcemaps.init())
+    .pipe(ts(project))
+    .pipe(sourcemaps.write())
+    .pipe(gulp.dest(_localFile('tmp'))); // Write every JS file to tmp/
+}, function() {
```
Comment (djrenren) [inline app/gulpfile.js:29]: Why are we writing the js files to disk at all? Gulp uses a virtual filesystem to track this stuff. Shouldn't we just pipe it into the next step at line 32?

File: Gulpfile.js (hunk 1 above)
Comment (bbesmanoff) [inline Gulpfile.js:9]: It does feel weird to call out that each project knows that it's a subproject.  When gulp registers the tasks, they are all added to the same registry (overwriting what is there).

File: app/gulpfile.js (hunk 4 above)
Comment (bbesmanoff) [inline app/gulpfile.js:29]: From what I can tell, it's required for webpack (it seems to read a file from disk).  There may be alternative libraries with support for reading from stream (found [`webpack-stream`](https://github.com/shama/webpack-stream)).

Comment (bbesmanoff) [inline app/gulpfile.js:29]: We are leaving this as it is due to how `gulp-webpack` works.

**Comment by djrenren** [comment]
lgtm

**Pull request merged**